src/gensxg.c

```diff
@@ -0,0 +1,377 @@
+} Options;
+
+static EVP_PKEY* load_private_key(const char* filepath) {
+  FILE* const keyfile = fopen(filepath, "r");
+  if (keyfile == NULL) {
+    return NULL;
+  }
+  EVP_PKEY* private_key =
+      PEM_read_PrivateKey(keyfile, NULL, NULL, NULL);
+  fclose(keyfile);
+  return private_key;
+}
+
+static X509* load_x509_cert(const char* filepath) {
+  FILE* certfile = fopen(filepath, "r");
+  if (certfile == NULL) {
+    return NULL;
+  }
+  X509* cert = PEM_read_X509(certfile, NULL, NULL, NULL);
+  fclose(certfile);
+  return cert;
+}
+
+static EVP_PKEY* load_ed25519_pubkey(const char* filepath) {
+  FILE* keyfile = fopen(filepath, "r");
+  if (keyfile == NULL) {
+    return NULL;
+  }
+  EVP_PKEY* public_key = PEM_read_PUBKEY(keyfile, NULL, NULL, NULL);
+  fclose(keyfile);
+  return public_key;
+}
+
+static uint64_t parse_time(const char datestring[]) {
+  struct tm date;
+  char* finished = strptime(datestring, "%Y-%m-%dT%H:%M:%SZ", &date);
+  if (finished == NULL) {
+    fprintf(stderr, "Failed to parse date: %s\n", datestring);
+  } else if(*finished != '\0') {
+    fprintf(stderr, "Failed to parse at: %s\n", finished);
+    exit(1);
+  }
+  return (uint64_t)mktime(&date);
+}
+
+Options parse_options(int argc, char* const argv[]) {
+  Options result;
+  memset(&result, 0, sizeof(result));
+  // Set default values.
+  result.help = false;
+  result.content_type = "text/html";
+  time_t now = time(NULL);
+  result.date = (uint64_t)now;
+  result.mi_record_size = 4096;
+
+  int opt;
+  int longindex;
+  while ((opt = getopt_long_only(argc,
+                                 argv,
+                                 "h:u:r:v:e:p:k:c:t:d:x:m:",
+                                 kOptions,
+                                 &longindex)) != -1) {
+    switch (opt) {
+      case 'h':
+        result.help = true;
+        return result;
+      case 'u':
+        result.url = optarg;
+        break;
+      case 'r':
+        result.cert_url = optarg;
+        break;
+      case 'v':
+        result.validity_url = optarg;
+        break;
+      case 'e':
+        result.certificate = optarg;
+        break;
+      case 'p':
+        result.public_key = optarg;
+        break;
+      case 'k':
+        result.private_key = optarg;
+        break;
+      case 'c':
+        result.content = optarg;
+        break;
+      case 't':
+        result.content_type = optarg;
+        break;
+      case 'd':
+        result.date = parse_time(optarg);
+        break;
+      case 'x':
+        result.expire = parse_time(optarg);
+        break;
+      case 'm':
+        result.mi_record_size = strtoull(optarg, NULL, 0);
+        break;
+      default:
+        exit(1);
+    }
+  }
+
+  if (result.expire == 0) {
+    result.expire = result.date + 60 * 60 * 24 * 7 - 1;  // 7 days.
+  }
+
+  return result;
+}
+
+static bool validate(const Options* opt) {
+  bool valid = true;
+  if (opt->url == NULL) {
+    fprintf(stderr, "error: -url must not be empty.\n");
```

Comment:
If you really want to reject the empty url, you should test `*opt->url == '\0'` (probably in addition to null-check you already do).